networking/eth2/src/main/java/tech/pegasys/artemis/networking/eth2/gossip/topics/validation/BlockValidator.java

```diff
@@ -0,0 +1,155 @@
+
+    final Optional<BeaconState> preState =
+        Optional.ofNullable(
+            recentChainData.getStore().getBlockState(block.getMessage().getParent_root()));
+    if (preState.isEmpty() || blockIsFromFutureSlot(block)) {
+      LOG.trace(
+          "BlockValidator: Either block pre state does not exist or block is from the future. "
+              + "It will be saved for future processing");
+      return BlockValidationResult.SAVED_FOR_FUTURE;
+    }
+
+    try {
+      BeaconState postState =
+          stateTransition.process_slots(preState.get(), block.getMessage().getSlot());
+      if (blockIsProposedByTheExpectedProposer(block, postState)
+          && blockSignatureIsValidWithRespectToProposerIndex(block, preState.get(), postState)) {
+        return BlockValidationResult.VALID;
+      }
+    } catch (EpochProcessingException | SlotProcessingException e) {
+      LOG.error("BlockValidator: Unable to process block state.", e);
+      return BlockValidationResult.INVALID;
+    }
+
+    return BlockValidationResult.INVALID;
+  }
+
+  private boolean blockIsFromFutureSlot(SignedBeaconBlock block) {
+    ReadOnlyStore store = recentChainData.getStore();
+    final long disparityInSeconds = Math.round((float) MAXIMUM_GOSSIP_CLOCK_DISPARITY / 1000.0);
+    final UnsignedLong maxOffset = UnsignedLong.valueOf(disparityInSeconds);
+    final UnsignedLong maxTime = store.getTime().plus(maxOffset);
+    UnsignedLong maxCurrSlot = getCurrentSlot(maxTime, store.getGenesisTime());
+    return block.getSlot().compareTo(maxCurrSlot) > 0;
+  }
+
+  private boolean blockSlotIsGreaterThanLatestFinalizedSlot(SignedBeaconBlock block) {
+    UnsignedLong finalizedSlot =
+        recentChainData.getStore().getFinalizedCheckpoint().getEpochStartSlot();
+    return block.getSlot().compareTo(finalizedSlot) > 0;
+  }
+
+  private boolean blockIsFirstBlockWithValidSignatureForSlot(SignedBeaconBlock block) {
+    return !receivedValidBlockInfoSet.contains(new SlotAndProposer(block));
+  }
+
+  private boolean blockSignatureIsValidWithRespectToProposerIndex(
+      SignedBeaconBlock block, BeaconState preState, BeaconState postState) {
+    Validator proposer =
+        postState
+            .getValidators()
+            .get(toIntExact(block.getMessage().getProposer_index().longValue()));
+
+    final Bytes domain = get_domain(preState, DOMAIN_BEACON_PROPOSER);
+    final Bytes signing_root = compute_signing_root(block.getMessage(), domain);
+    final BLSSignature signature = block.getSignature();
+    boolean signatureValid = BLS.verify(proposer.getPubkey(), signing_root, signature);
+
+    return signatureValid && receivedValidBlockInfoSet.add(new SlotAndProposer(block));
+  }
+
+  private boolean blockIsProposedByTheExpectedProposer(
+      SignedBeaconBlock block, BeaconState postState) {
+    final int proposerIndex = get_beacon_proposer_index(postState);
+    return proposerIndex == toIntExact(block.getMessage().getProposer_index().longValue());
```

Comment:
Remove the `toIntExact` here and use long comparison (the int proposerIndex will be automatically up-cast to a long).  That way it will evaluate to false if proposer index is greater than Integer.MAX_VALUE rather than throwing an exception.
```suggestion
    return proposerIndex == block.getMessage().getProposer_index().longValue();
```

Reply:
done.